### DISCUSSION TITLE
ISR + Subdomain Wildcards

### DISCUSSION THREAD

**Discussion by Utkarshbhimte** [vercel/next.js#25327] [Help] [answered]
### Describe the feature you'd like to request

There is no way currently to implement static caching on a website that handles multiple subdomains, 

I am making a website that would help users to host their own microsite and link it with their own domain, Though the domain linking is possible using Vercel wildcards, 

I am aiming to improve the performance of these pages which will be hosted on custom domains. 

### Describe the solution you'd like

I am currently exploring the Nextjs codebase but as far as I can understand, 
the ISR caching is happening on the page level, and you can generate multiple pages using `getStaticPaths`, 

I think we can include the wildcard in the `getStaticPaths` as a parameter. 

```js
export async function getStaticPaths() {
  return {
    paths: [
      { params: { ... , wildcard: '' } } 
    ],
    fallback: true or false 
  };
}
```

and this wildcard can show up in `getStaticProps` as a key. 

This would enable devs to build no-code tools that will provide robust and performant microsites. This would also be helpful for websites that have localization on the subdomain level to have ISR without running multiple servers/projects.

### Describe alternatives you've considered

The other approach here would be to build the app statically for each person and upload it to a storage bucket and link that user-provided domain. but this won't have ISR goodness. 

Another approach could be to have dynamic caching using a service like cloudflare.

**Comment by lalit2005** [comment]
Were you able to find any solution?

**Reply by steven-tey** [reply]
@Lalit2005 I just posted an example for this in my comment below, hope it helps!

**Comment by steven-tey** [comment]
Hey @Utkarshbhimte! We recently rolled out a [simple example](https://github.com/vercel/examples/tree/main/edge-functions/hostname-rewrites) that allows you to do exactly this – create unique content pages with a multi-tenant infrastructure using [edge function](https://vercel.com/features/edge-functions) rewrites and [ISR](https://vercel.com/docs/concepts/next.js/incremental-static-regeneration). Each of your users gets assigned a unique subdomain when they create their account, with the (usually paid) option to add a custom domain. 

We're also rolling out more complex boilerplates (e-commerce storefront builders, website builders) + supplementary code snippets (static tweets, domains API, OG image generator) in an [early access program](https://eap.vercel.sh/) – feel free to join if you're interested (click "sign up" on the Platforms on Vercel program).

**Comment by steven-tey** [comment, marked as answer]
@Utkarshbhimte  Quick update on this – we recently launched the Platforms Starter Kit that's a comprehensive template for building multi-tenant applications with built-in custom domains support.: https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Reply by omsharp** [reply]
Hi @steven-tey,

I'm using your example to build a multi-tenant application, and its working great. However, whenever I use next Link to navigate to another page, the application does a hard page refresh instead of the normal Link behavior.

When I link directly to _sites/[site]/page it works fine. It only hard refreshes when I rewrite to the user's subdomain url in the middleware.

Is that a normal behavior?


[UPDATE] 
Problem solved by updating Next to 12.1 , I was using 12.0.7 when I had the issue.
Not sure what the problem was, but glad it's solved anyway.

**Reply by farrokh** [reply]
Hi @steven-tey . Thanks for such a great kit. I'm using your middleware settings on my self-hosted SAAS project. Everything is working great. However, I have an issue with deep links. your middleware setting can easily map `example.example.com/yyy` to `www.example.com/_sites/example/yyy` but the issue raises when I try to go deeper. For example `example.example.com/yyy/ddd/lll/` . When I do that it returns 404!

here is the folder structure I have:

```
pages
-_sites
--[id]
---yyy
----ddd
-----lll.js
---[slug].js
```
is there a way to handle deep paths too?

**Reply by farrokh** [reply]
Nevermind I figured out I was missing `getStaticPaths` in my index file:
```
export const getStaticPaths = async () => {
  const allPaths = ["xyz"];
  return {
    paths: allPaths.map((path) => ({
      params: {
        site: path,
      },
    })),
    fallback: true,
  };
};
```

Thanks again for such great work.

**Reply by steven-tey** [reply]
@omsharp @farrokh sorry for the late reply guys, but glad to see that y'all were able to make it work! LFG 🚀
